Significant Eso fixes and improvements

jwoillez opened the pull request:
This PR does the following:
- Add destination keyword to Eso.retrieve_data(), to download files to a specific location (other than the cache).
- Fixed Eso.query_instrument() to use instrument specific query forms (it was using the main form before).
- Implemented Eso.query_main() to query all instruments with the main form (even the ones without a specific form).
- Disabled caching for all Eso.retrieve_data() operations (that did not make sense).

This PR comes with:
- documentation updates (for the above changes) and fixes.
- tests update.

Todo:
- [x] Fix the remote test `test_SgrAstar_remotevslocal` (<- I don't understand what it does. Help needed)

pep8speaks commented:
Hello @jwoillez! Thanks for submitting the PR.

 - In the file [`astroquery/eso/core.py`](https://github.com/astropy/astroquery/blob/d396beebc27c37b14e5a9ce49cd27c3a64390df1/astroquery/eso/core.py), following are the PEP8 issues :

> [Line 378:21](https://github.com/astropy/astroquery/blob/d396beebc27c37b14e5a9ce49cd27c3a64390df1/astroquery/eso/core.py#L378): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 409:29](https://github.com/astropy/astroquery/blob/d396beebc27c37b14e5a9ce49cd27c3a64390df1/astroquery/eso/core.py#L409): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 447:29](https://github.com/astropy/astroquery/blob/d396beebc27c37b14e5a9ce49cd27c3a64390df1/astroquery/eso/core.py#L447): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 450:17](https://github.com/astropy/astroquery/blob/d396beebc27c37b14e5a9ce49cd27c3a64390df1/astroquery/eso/core.py#L450): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent

jwoillez commented:
This PR should be ready for review. I would also need some information on the purpose of the `test_SgrAstar_remotevslocal` remote test (the only that fails on my machine).

jwoillez commented:
Note: the Tracis-CI failure is not related to this PR.
`astropy.utils.exceptions.AstropyDeprecationWarning: astropy.utils.compat.futures is now deprecated - use concurrent.futures instead`

jwoillez commented:
I added another small fix that improves the help message returned by e.g. `Eso.query_instrument('gravity', help=True)`. Some of the available options were missing, being wrapped in paragraphs (`<p> ... </p>`).

keflavich commented:
@jwoillez thanks for all these fixes.  I will review when I have a chance.  We probably need to update the travis build matrix to handle the errors you're seeing.

keflavich commented:
Thanks again @jwoillez.  I don't see any problems with the changes you've made; this PR looks like a great improvement to the ESO archive interface.  I think merging or rebasing will get the tests to pass, since that will bring in the appropriate changes to travis-ci.

jwoillez commented:
Thanks. I will rebase later tonight.

jwoillez commented:
@keflavich I just rebased. Let's see how it goes. I still need your help understanding the purpose of the following remote test that currently fails for me: `test_SgrAstar_remotevslocal`

keflavich commented:
@jwoillez that's a check to make sure we're getting consistent results between the actual query to the server and with the local data we've cached.  If that's failing, it implies we need to update the local test data.

bsipocz commented:
@jwoillez @keflavich - I'm looking into that, I'm fairly certain that it should be unrelated to this PR.

jwoillez commented:
Rebased again, and fixed the remote test that was failing. Can someone check that the test is as intended? It now compares `query_instrument(cache=False)` to `query_instrument(cache=True)`.

bsipocz commented:
Yes, that cache False/True was indeed one of the issues, but I still don't see why it doesn't return any results. Maybe the instrument list evolved over the months/years, and while it had results for one of the instruments, there is none for 'fors1' now.

bsipocz commented:
Oh, OK. The issue was indeed that the ``instruments[0]`` was constantly changing based on the order the server returned the instruments. So your fix should indeed work.

bsipocz commented:
@jwoillez @keflavich - One issue, that's only somewhat related to this issue. Namely is that I think we should get rid of the pickle files and store that data in a different format as pickles should not be used for long term data storage...
